src/puppetlabs/puppetdb/cli/time_shift_export.clj

```diff
@@ -4,32 +4,174 @@
+              (update :events (fn [events]
+                                (mapv (fn [event]
+                                        (update event :timestamp #(time-shift-timestamp % shift-period)))
+                                      events)))))
+        resources))
+
+(defn time-shift-logs
+  [logs shift-period]
+  (mapv (fn [log]
+          (update log :time #(time-shift-timestamp % shift-period)))
+        logs))
+
+(defn time-shift-producer-timestamp
+  [json-data shift-period]
+  (update json-data :producer_timestamp #(time-shift-timestamp % shift-period)))
+
+(defn time-shift-report-timestamps
+  [json-data shift-period]
+  (-> json-data
+      (time-shift-producer-timestamp shift-period)
+      (update :start_time #(time-shift-timestamp % shift-period))
+      (update :end_time #(time-shift-timestamp % shift-period))
+      (update :logs #(time-shift-logs % shift-period))
+      (update :resources #(time-shift-resources % shift-period))))
+
+
+(defn time-shift-timestamps-from-entry
+  [shift-period tar-reader tar-entry]
+  (let [json-data (-> tar-reader
+                      archive/read-entry-content
+                      (json/parse-string true))]
+    (condp #(str/includes? %2 %1) (.getName tar-entry)
+      "catalogs" (time-shift-producer-timestamp json-data shift-period)
+      "reports" (time-shift-report-timestamps json-data shift-period)
+      "facts" (time-shift-producer-timestamp json-data shift-period)
+      "export-metadata" json-data)))
+
+(defn time-shift-input-archive
+  [input-path output-path shift-period]
+  (with-open [tar-writer (archive/tarball-writer (or output-path System/out))]
+    (with-open [tar-reader (archive/tarball-reader input-path)]
+      (doseq [tar-entry (archive/all-entries tar-reader)]
+        (->> tar-entry
+             (time-shift-timestamps-from-entry shift-period tar-reader)
+             json/generate-pretty-string
+             (archive/add-entry tar-writer "utf-8" (.getName tar-entry)))))))
+
+(defn get-newest-timestamp
+  [first-timestamp second-timestamp]
+  (if (or (nil? second-timestamp)
+          (after? first-timestamp second-timestamp))
+    first-timestamp
+    second-timestamp))
+
+(defn resources-newest-timestamp
+  [resources]
+  (->> (mapcat (fn [resource]
+                 (concat [(parse-wire-datetime (resource :timestamp))]
+                         (map #(parse-wire-datetime (:timestamp %)) (resource :events))))
+               resources)
+       (reduce get-newest-timestamp)))
+
+(defn newest-timestamp-from-report
+  [json-data]
+  (->> (concat [(parse-wire-datetime (json-data :producer_timestamp))
+                (parse-wire-datetime (json-data :start_time))
+                (parse-wire-datetime (json-data :end_time))]
+               (map #(parse-wire-datetime (:time %)) (json-data :logs)))
+       (reduce get-newest-timestamp)
+       (get-newest-timestamp (resources-newest-timestamp (json-data :resources)))))
```

Comment:
The current version might be faster, but I think this 
`(defn get-resources-timestamps
  [resources]
  (mapcat (fn [resource]
            (concat [(resource :timestamp)]
                    (map #(:timestamp %) (resource :events))))
          resources))

(defn newest-timestamp-from-report
  [json-data]
  (->> (concat [(json-data :producer_timestamp)
                (json-data :start_time)
                (json-data :end_time)]
               (map #(:time %) (json-data :logs))
               (get-resources-timestamps (json-data :resources)))
       (map parse-wire-datetime)
       (reduce get-newest-timestamp)))` 
looks better